src/H5Sselect.c

```diff
@@ -218,26 +218,39 @@ H5Sselect_copy(hid_t dst_id, hid_t src_id)
 herr_t
 H5S_select_copy(H5S_t *dst, const H5S_t *src, bool share_selection)
 {
-    herr_t ret_value = FAIL; /* Return value */
+    H5S_t  tmp_space;
+    bool   copied_space = false;
+    herr_t ret_value    = FAIL; /* Return value */
 
     FUNC_ENTER_NOAPI(FAIL)
 
     /* Check args */
     assert(dst);
     assert(src);
 
-    /* Release the current selection */
-    if (H5S_SELECT_RELEASE(dst) < 0)
```

Comment:
Part of this PR was fixing `H5Scopy()` to return failure for datasets with a NULL extent (rank value of 0). However, the previous code didn't really account for failure and would release the selection in the destination space before determining if the selection copy portion actually succeeded or not. This could result in segfaults later on when the application tries to close the destination dataspace. This reworks things so the copy is made into a temporary dataspace and only releases the selection in the destination dataspace once things have succeeded.